### ISSUE TITLE
panic out of range in bytes queue

### ISSUE THREAD

**Issue by codyohl** [allegro/bigcache#148]
https://github.com/allegro/bigcache/blob/bbf64ae21fc5555f4e9752825ee9ffe13b1e5fa0/queue/bytes_queue.go#L222

It appears there is a bounds check before the blockSize, but then there is an assumption that adding block size is not out of bounds.

**Comment by cristaloleg** [comment]
Hi, can you add a bit more details, please? What is the problem?

**Comment by codyohlsen** [comment]
@cristaloleg, I can dig a little deeper into the bytes_queue to figure out what might be the right fix, but the underlying issue is that 
https://github.com/allegro/bigcache/blob/bbf64ae21fc5555f4e9752825ee9ffe13b1e5fa0/queue/bytes_queue.go#L223
results in a panic because` index+headerEntrySize+blockSize  >=len(q.array)` .

There is a bounds check on L218 here: https://github.com/allegro/bigcache/blob/bbf64ae21fc5555f4e9752825ee9ffe13b1e5fa0/queue/bytes_queue.go#L218

but there is no bounds check for adding the blockSize to that. Does that make sense?

**Comment by cristaloleg** [comment]
Oh, finally I got it. I thought there is a problem on the line 222 as in the original message.

That's funny that there were no(???) panic before. I think somewhere in late 2017 I've checked this code and everything was fine. I might be wrong 😉

**Comment by codyohlsen** [comment]
@cristaloleg it might be an uncommon edge case, or a misuse of the API somehow, but I did see this panic in logs using this code. If I have time, I can try to find a small reproducible test, it but it might be tough without enough knowledge about this code. I do experience this panic without changing this library, so some permutation of the usage of this API will cause the panic

**Comment by olivierperes** [comment]
@cristaloleg I obtain this panic while using an iterator to check all entries whose name match a given pattern and deleting them, while other processes are adding entries. It is also possible that two of these iterators work concurrently, but they cannot try to delete the same entries. Is this a supported pattern?

**Comment by janisz** [comment]
@olivierperes Can you provide code to demonstrate this behavior?

**Comment by olivierperes** [comment]
This is the function that deletes entries. Unfortunately I cannot provide a unit test. My program can run for about 12 hours before the crash occurs.
```go
// cleanCache deletes from the cache all elements whose name has the given prefix.
func cleanCache(prefix string) error {
	iterator := cache.Iterator()
	for iterator.SetNext() {
		entry, err := iterator.Value()
		if err != nil {
			return err
		}
		k := entry.Key()
		if strings.HasPrefix(k, prefix) {
			err := cache.Delete(k)
			if err != nil && err != bigcache.ErrEntryNotFound {
				return err
			}
		}
	}
	return nil
}
```

**Comment by janisz** [comment]
> My program can run for about 12 hours before the crash occurs.
That's important information. I suspect your program alocates more and more memory until it reaches 

https://github.com/allegro/bigcache/blob/c00e2be7ad487b8fe06fc1da3d62edd19fb54519/config.go#L23-L27

So here 
https://github.com/allegro/bigcache/blob/e24eb225f15679bbe54f91bfa7da3b00e59b9768/queue/bytes_queue.go#L81

we may allocate less memory than we need

I think the bug may be here 
https://github.com/allegro/bigcache/blob/e24eb225f15679bbe54f91bfa7da3b00e59b9768/queue/bytes_queue.go#L134-L136

We do not check if all data were succesfuly copied

**Comment by janisz** [comment]
Here is small visualization https://storytime.dev/preview/5d8b85caddbb4e10b4566fc9

**Comment by olivierperes** [comment]
Hello @janisz. Just to be clear, is someone working on this or are you expecting a PR? In principle I would volunteer but in this case, it seems to require a very good knowledge of the code.

**Comment by janisz** [comment]
I tired to reproduce this issue with approach I presented above but I can't do it. It's hard for me to figure out the order for inseritions/deletions to get invalid state.

**Comment by janisz** [comment]
How about using https://github.com/dvyukov/go-fuzz to reproduce this bug. I came up with following   piece of code. I have no experience in fuzzing so it might be totally wrong but at least we can iterate on this. We know bug is in [Queue](/queue.go) and my approach is to use random data to perform random operations on Queue (Push with random length and Pop).

```go
package fuzz

import (
	"bytes"
	"encoding/binary"
	"io"

	. "github.com/allegro/bigcache/queue"
)

func Fuzz(data []byte) int {

	queue := NewBytesQueue(0, 4 * 1024 * 1024 * 1024, false)

	r := bytes.NewReader(data)
	for {
		push, n, err := nextMove(r)
		if err == io.EOF {
			return 1
		}
		if err != nil {
			panic(err)
		}
		if push {
			queue.Push(blob('a', int(n)))
		} else {
			queue.Pop()
		}
	}

	return 1
}

func nextMove(r io.Reader) (push bool, n int16, err error) {
	boolean := make([]byte, 1)
	_, err = r.Read(boolean)
	if err != nil {
		return false, n, err
	}
	if boolean[0] == 0 {
		return false, n, nil
	}
	integer := make([]byte, 16)
	_, err = r.Read(integer)
	if err != nil {
		return false, n, err
	}
	err = binary.Read(bytes.NewReader(integer), binary.BigEndian, &n)
	return true, n, err
}

func blob(char byte, len int) []byte {
	b := make([]byte, len)
	for index := range b {
		b[index] = char
	}
	return b
}
```

**Comment by janisz** [comment]
This is similar to: https://github.com/janisz/bigcache/commit/e73c96d8b9c30e2f5247ce96521edbd62b5d096b/checks
```
goroutine 153 [running]:
github.com/allegro/bigcache/v2.readKeyFromEntry(0xc0287db076, 0x41b, 0x4f8a, 0xec9cb036dfa3401a, 0xc0287db076)
	/home/runner/work/bigcache/bigcache/encoding.go:51 +0x211
github.com/allegro/bigcache/v2.(*cacheShard).get(0xc013c59680, 0xc05661abf0, 0x9, 0xec9cb036dfa3401a, 0x0, 0x0, 0xc000165ec8, 0x512507, 0xc05661f680)
	/home/runner/work/bigcache/bigcache/shard.go:69 +0x120
github.com/allegro/bigcache/v2.(*BigCache).Get(0xc00015c000, 0xc05661abf0, 0x9, 0x1, 0x1, 0xc05661abf0, 0x9, 0x0)
	/home/runner/work/bigcache/bigcache/bigcache.go:118 +0x175
github.com/allegro/bigcache/v2.TestCacheDelRandomly.func3(0xc3500, 0xc00015c000, 0xc000110400, 0xc01429e5e0)
	/home/runner/work/bigcache/bigcache/bigcache_test.go:493 +0xd9
created by github.com/allegro/bigcache/v2.TestCacheDelRandomly
	/home/runner/work/bigcache/bigcache/bigcache_test.go:484 +0x264
exit status 2
```

**Comment by mxplusb** [comment]
> I have no experience in fuzzing so it might be totally wrong but at least we can iterate on this.

I'm also not an expert, but you can prolly come up with a test case like you have already and then just loop over it N times with a fuzzer providing some data and then when it crashes, we'll see what happens.

/shrug

**Comment by janisz** [comment]
```diff
@@ -46,6 +48,10 @@ func readTimestampFromEntry(data []byte) uint64 {
 func readKeyFromEntry(data []byte) string {
        length := binary.LittleEndian.Uint16(data[timestampSizeInBytes+hashSizeInBytes:])
 
+       if headersSizeInBytes+int(length) > len(data) {
+               panic(fmt.Sprintf("\n%s\n", hex.Dump(data)))
+       }
+
        // copy on read
        dst := make([]byte, length)
        copy(dst, data[headersSizeInBytes:headersSizeInBytes+length])
```

```
go test -run TestCacheDelRandomly -v -count 100
=== RUN   TestCacheDelRandomly
=== PAUSE TestCacheDelRandomly
=== CONT  TestCacheDelRandomly
2019/12/02 19:20:42 Allocated new queue in 1.179µs; Capacity: 2308 
2019/12/02 19:20:42 Allocated new queue in 10.236µs; Capacity: 4616 
2019/12/02 19:20:42 Allocated new queue in 9.476µs; Capacity: 9232 
2019/12/02 19:20:42 Allocated new queue in 18.23µs; Capacity: 18464 
2019/12/02 19:20:42 Allocated new queue in 43.117µs; Capacity: 36928 
2019/12/02 19:20:42 Allocated new queue in 37.135µs; Capacity: 73856 
2019/12/02 19:20:42 Allocated new queue in 90.396µs; Capacity: 147712 
2019/12/02 19:20:42 Allocated new queue in 113.411µs; Capacity: 295424 
2019/12/02 19:20:42 Allocated new queue in 191.364µs; Capacity: 590848 
2019/12/02 19:20:42 Allocated new queue in 351.637µs; Capacity: 1048576 
panic: 
00000000  6e 6e 6e 6e 6e 6e 6e 6e  6e 6e 6e 6e 6e 6e 6e 6e  |nnnnnnnnnnnnnnnn|
00000010  6e 6e 6e 6e 65 6b 65 79  31 38 34 b8 b8 b8 b8 b8  |nnnnekey184.....|
00000020  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000030  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000040  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000050  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000060  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000070  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000080  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000090  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000000a0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000000b0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000000c0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000000d0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000000e0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000000f0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000100  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000110  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000120  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000130  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000140  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000150  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000160  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000170  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000180  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000190  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000001a0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000001b0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000001c0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000001d0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000001e0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000001f0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000200  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000210  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000220  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000230  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000240  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000250  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000260  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000270  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000280  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000290  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000002a0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000002b0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000002c0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000002d0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000002e0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000002f0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000300  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000310  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000320  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000330  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000340  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000350  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000360  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000370  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000380  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000390  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000003a0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000003b0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000003c0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000003d0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000003e0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
000003f0  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000400  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8 b8 b8 b8 b8 b8  |................|
00000410  b8 b8 b8 b8 b8 b8 b8 b8  b8 b8 b8                 |...........|



goroutine 23 [running]:
github.com/allegro/bigcache/v2.readKeyFromEntry(0xc000343c32, 0x41b, 0x643ce, 0xec95a236df9d09a2, 0xc000343c32)
	/home/janisz/go/src/github.com/allegro/bigcache/encoding.go:52 +0x231
github.com/allegro/bigcache/v2.(*cacheShard).get(0xc0000f2000, 0xc00044c580, 0x9, 0xec95a236df9d09a2, 0xc00044c580, 0x9, 0x9, 0xc00044c580, 0xc0000b20d0)
	/home/janisz/go/src/github.com/allegro/bigcache/shard.go:69 +0xcd
github.com/allegro/bigcache/v2.(*BigCache).Get(0xc000092ea0, 0xc00044c580, 0x9, 0x1, 0x1, 0xc00044c580, 0x9, 0x0)
	/home/janisz/go/src/github.com/allegro/bigcache/bigcache.go:118 +0x8b
github.com/allegro/bigcache/v2.TestCacheDelRandomly.func3(0xc3500, 0xc000092ea0, 0xc0000dc100, 0xc0000a0110)
	/home/janisz/go/src/github.com/allegro/bigcache/bigcache_test.go:491 +0x96
created by github.com/allegro/bigcache/v2.TestCacheDelRandomly
	/home/janisz/go/src/github.com/allegro/bigcache/bigcache_test.go:482 +0x294
exit status 2
FAIL	github.com/allegro/bigcache/v2	1.059s

```
It looks like two entries can override each other.

```
go version
go version go1.13.4 linux/amd64
```

**Comment by janisz** [comment]
https://storytime.dev/preview/5de55ab8ddbb4e10b45671df

**Comment by mxplusb** [comment]
> It looks like two entries can override each other.

Race condition or bad locking?

**Comment by janisz** [comment]
I think it could be a bad locking
@codyohlsen Do you experience same problems with the latest version of bigcache?

**Comment by mxplusb** [comment]
I saw this before the PR, and I would agree it's likely bad locking. Ideally, bigcache should deterministically lock on any operation, but I'm not really sure if that's reasonable.

**Comment by janisz** [comment]
This just hit me 
```
panic: runtime error: slice bounds out of range

goroutine 857 [running]:
panic(0x510f40, 0xc4200100c0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/allegro/bigcache.(*cacheShard).get(0xc51e950360, 0xc4c8342220, 0x8, 0x20ab7654471937b6, 0x4b84c4, 0x10fdd43, 0xa, 0xc4c8342220, 0x8)
	/go/src/github.com/allegro/bigcache/shard.go:72 +0x68c
github.com/allegro/bigcache.(*BigCache).Get(0xc43579e1a0, 0xc4c8342220, 0x8, 0xc4c8353e00, 0x100, 0x100, 0x0, 0x0)
	/go/src/github.com/allegro/bigcache/bigcache.go:118 +0x92
github.com/allegro/bigcache.readFromCache.func1(0xc4df9d0000)
	/go/src/github.com/allegro/bigcache/bigcache_bench_test.go:160 +0x101
testing.(*B).RunParallel.func1(0xc4c82c8720, 0xc4c82c8718, 0xc4c82c8710, 0xc4fdfbc280, 0xc51e9520a0)
	/usr/local/go/src/testing/benchmark.go:598 +0xc4
created by testing.(*B).RunParallel
	/usr/local/go/src/testing/benchmark.go:599 +0x1a2
```
On `BenchmarkReadFromCacheWithInfo/1-shards-4                    	signal: interrupt`
Using Go go1.14beta1_linux_amd64 test was slightly modified I set `HardMaxCacheSize:   16 * 1024,`

**Comment by janisz** [comment]
I prepared a test for this issue – unfortunately it's far from minimal but still reproducible example 
```go
func Test_issue_148(t *testing.T) {
	const n = 9000000
	var message = blob('a', 512)
	cache, _ := NewBigCache(Config{
		Shards:             1,
		LifeWindow:         time.Hour,
		MaxEntriesInWindow: 10,
		MaxEntrySize:       len(message) + 1,
		HardMaxCacheSize:   n >> 11,
	})
	for i := 0; i < n; i++ {
		err := cache.Set(strconv.Itoa(i), message)
		if err != nil {
			t.Fatal(err)
		}
	}

	wg := sync.WaitGroup{}
	wg.Add(n)
	for i := 0; i < n; i++ {
		go func() {
			cache.Get(strconv.Itoa(rand.Intn(n)))
			wg.Done()
		}()
	}
	wg.Wait()
}
```

**Comment by mxplusb** [comment]
The fact this feels like a transient locking issue is strange. I might try to take some time this week to work on it since I have some cycles.

**Comment by janisz** [comment]
I found 3 problems:
* Peek can return empty data without an error (this will cause panic when trying to decode entry)
https://github.com/allegro/bigcache/blob/32b9aa9b686b8fc148dcb98eb31e830fe58d5982/queue/bytes_queue.go#L208-L216
* We do not check if initial capacity is lower than hard max shard size
https://github.com/allegro/bigcache/blob/32b9aa9b686b8fc148dcb98eb31e830fe58d5982/queue/bytes_queue.go#L47-L59
* length + index can be bigger than `len(data)`
https://github.com/allegro/bigcache/blob/0bde97d3259d05932b6376e38370413559b5385d/encoding.go#L46-L54

**Comment by cristaloleg** [comment]
@janisz what is the status of the issue?

**Comment by janisz** [comment]
We still need to fix it

**Comment by mxplusb** [comment]
I think #232 tentatively fixes this. I want to keep this open for a bit to see if we're still running across it.

@olivierperes @codyohlsen can you reproduce the issue on the latest commit?

**Comment by olivierperes** [comment]
> can you reproduce the issue on the latest commit?

I changed job and no longer have access to the code that used BigCache, so I won’t be able to test.

**Comment by janisz** [comment]
@codyohl @cristaloleg  can we close it. It was fixed by #236

**Comment by codyohl** [comment]
@janisz Yes, i don't observe this anymore. I'm all good with closing!
